File: agent/app/src/main/kotlin/dev/pilot/agent/SocketServer.kt
package dev.pilot.agent

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.util.concurrent.Executors

/**
 * TCP socket server that listens for JSON commands from the host daemon.
 *
 * Protocol: newline-delimited JSON. Each line is a complete JSON object.
 * Request:  {"id": "uuid", "method": "methodName", "params": {...}}
 * Response: {"id": "uuid", "result": {...}} or {"id": "uuid", "error": {...}}
 *
 * Commands are dispatched to the CommandHandler which runs UIAutomator2
 * operations. Each client connection is handled on a dedicated thread from
 * a fixed thread pool to ensure UIAutomator calls have the right context.
 */
class SocketServer(
    private val port: Int,
    private val commandHandler: CommandHandler,
) {
    companion object {
        private const val TAG = "PilotSocket"
    }

    private val executor = Executors.newFixedThreadPool(2)
    private var serverSocket: ServerSocket? = null

    @Volatile
    private var running = false

    suspend fun start() {
        running = true
        withContext(Dispatchers.IO) {
            try {
                serverSocket = ServerSocket(port)
                Log.i(TAG, "Listening on port $port")

                while (running) {
                    val client =
                        try {
                            serverSocket?.accept()
                        } catch (e: SocketException) {
                            if (running) Log.e(TAG, "Accept failed", e)
                            break
                        } ?: break

                    Log.i(TAG, "Client connected: ${client.remoteSocketAddress}")
                    // Handle each client on a worker thread so UIAutomator
                    // operations run with the correct thread context
                    executor.submit { handleClient(client) }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Server error", e)
            } finally {
                Log.i(TAG, "Server stopped")
            }
        }
    }

    fun stop() {
        running = false
        try {
            serverSocket?.close()
        } catch (_: Exception) {
        }
        executor.shutdownNow()
    }

    private fun handleClient(socket: Socket) {
        try {
            socket.use { s ->
                s.tcpNoDelay = true
                val reader = BufferedReader(InputStreamReader(s.getInputStream(), Charsets.UTF_8))
                val writer = PrintWriter(s.getOutputStream(), true)

                while (running && !s.isClosed) {
                    val line =
                        try {
                            reader.readLine()
                        } catch (e: SocketException) {
                            Log.d(TAG, "Client read error: ${e.message}")
                            null
                        }

                    if (line == null) {
                        Log.i(TAG, "Client disconnected")
                        break
                    }

                    if (line.isBlank()) continue

                    Log.d(TAG, "Received: $line")

                    val response =
                        try {
                            commandHandler.handle(line)
                        } catch (e: Exception) {
                            Log.e(TAG, "Unhandled error processing command", e)
                            val msg = e.message?.replace("\"", "\\\"") ?: "Unknown error"
                            """{"id":null,"error":{"type":"INTERNAL_ERROR","message":"$msg"}}"""
                        }

                    Log.d(TAG, "Responding: $response")

                    try {
                        writer.println(response)
                        writer.flush()
                    } catch (e: SocketException) {
                        Log.d(TAG, "Client write error: ${e.message}")
                        break
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Client handler error", e)
        }
    }
}

File: ios-agent/PilotAgent/SocketServer.swift
import Foundation
import Network

/// TCP socket server that listens for JSON commands from the host daemon.
///
/// Protocol: newline-delimited JSON. Each line is a complete JSON object.
/// Request:  {"id": "uuid", "method": "methodName", "params": {...}}
/// Response: {"id": "uuid", "result": {...}} or {"id": "uuid", "error": {...}}
///
/// Commands are processed on the main thread via RunLoop to ensure XCUITest's
/// accessibility operations (which require main RunLoop XPC callbacks) work
/// correctly on Xcode 26.
class SocketServer {
    private let port: UInt16
    private let commandHandler: CommandHandler
    private var listener: NWListener?
    private let queue = DispatchQueue(label: "dev.pilot.agent.socket", qos: .userInitiated)

    /// Pending commands from the socket, to be processed on the main thread.
    private var pendingCommands: [(String, NWConnection)] = []
    private let commandLock = NSLock()

    @Volatile private var running = false

    init(port: UInt16, commandHandler: CommandHandler) {
        self.port = port
        self.commandHandler = commandHandler
    }

    /// Start the server. This blocks the calling thread (main/test thread)
    /// using RunLoop so XCUITest's XPC callbacks can be processed.
    func start() {
        running = true

        let params = NWParameters.tcp
        params.allowLocalEndpointReuse = true

        // Disable Nagle's algorithm for low latency
        if let tcpOptions = params.defaultProtocolStack.internetProtocol as? NWProtocolTCP.Options {
            tcpOptions.noDelay = true
        }

        do {
            listener = try NWListener(using: params, on: NWEndpoint.Port(rawValue: port)!)
        } catch {
            NSLog("[PilotSocket] Failed to create listener: \(error)")
            return
        }

        listener?.newConnectionHandler = { [weak self] connection in
            guard let self = self, self.running else {
                connection.cancel()
                return
            }
            NSLog("[PilotSocket] Client connected")
            self.handleConnection(connection)
        }

        listener?.stateUpdateHandler = { state in
            switch state {
            case .ready:
                NSLog("[PilotSocket] Listening on port \(self.port)")
            case .failed(let error):
                NSLog("[PilotSocket] Listener failed: \(error)")
            case .cancelled:
                NSLog("[PilotSocket] Listener cancelled")
            default:
                break
            }
        }

        listener?.start(queue: queue)

        // Main loop: pump the RunLoop and process any pending commands.
        // XCUITest's accessibility operations require the main RunLoop to
        // process XPC callbacks. We process commands HERE on the main thread
        // rather than on the socket's background queue.
        while running {
            // Process one pending command if available
            commandLock.lock()
            let pending = pendingCommands.isEmpty ? nil : pendingCommands.removeFirst()
            commandLock.unlock()

            if let (line, connection) = pending {
                let response = commandHandler.handle(rawJson: line)
                let responseData = (response + "\n").data(using: .utf8)!
                connection.send(content: responseData, completion: .contentProcessed { error in
                    if let error = error {
                        NSLog("[PilotSocket] Write error: \(error)")
                        connection.cancel()
                    }
                })
            }

            // Run the RunLoop briefly to process XPC callbacks and GCD events.
            // Short interval when commands are pending, longer when idle.
            let interval = pending != nil ? 0.001 : 0.05
            RunLoop.current.run(mode: .default, before: Date(timeIntervalSinceNow: interval))
        }
    }

    /// Stop the server.
    func stop() {
        running = false
        listener?.cancel()
        listener = nil
    }

    private func handleConnection(_ connection: NWConnection) {
        connection.start(queue: queue)
        receiveLines(connection: connection, buffer: Data())
    }

    /// Recursively receive data and enqueue complete commands for main-thread processing.
    private func receiveLines(connection: NWConnection, buffer: Data) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 65536) { [weak self] data, _, isComplete, error in
            guard let self = self, self.running else {
                connection.cancel()
                return
            }

            if let error = error {
                NSLog("[PilotSocket] Read error: \(error)")
                connection.cancel()
                return
            }

            var currentBuffer = buffer
            if let data = data {
                currentBuffer.append(data)
            }

            // Process all complete lines in the buffer
            while let newlineIndex = currentBuffer.firstIndex(of: UInt8(ascii: "\n")) {
                let lineData = currentBuffer[currentBuffer.startIndex..<newlineIndex]
                currentBuffer = Data(currentBuffer[currentBuffer.index(after: newlineIndex)...])

                guard let line = String(data: lineData, encoding: .utf8), !line.trimmingCharacters(in: .whitespaces).isEmpty else {
                    continue
                }

                NSLog("[PilotSocket] Received: \(line.prefix(200))")

                // Enqueue for main-thread processing
                self.commandLock.lock()
                self.pendingCommands.append((line, connection))
                self.commandLock.unlock()
            }

            if isComplete {
                NSLog("[PilotSocket] Client disconnected")
                connection.cancel()
                return
            }

            // Continue reading
            self.receiveLines(connection: connection, buffer: currentBuffer)
        }
    }
}

// MARK: - Volatile property wrapper (equivalent to Kotlin's @Volatile)

@propertyWrapper
struct Volatile<Value> {
    private var value: Value
    private let lock = NSLock()

    init(wrappedValue: Value) {
        self.value = wrappedValue
    }

    var wrappedValue: Value {
        get {
            lock.lock()
            defer { lock.unlock() }
            return value
        }
        set {
            lock.lock()
            defer { lock.unlock() }
            value = newValue
        }
    }
}
